### DISCUSSION TITLE
Checking if a command has succeeded

### DISCUSSION THREAD

**Discussion by tueda** [mg5amcnlo/mg5amcnlo#128] [Q&A] [answered]
Is there an easy way to programmatically check if a command has succeeded or not?

Unfortunately, in shell scripts,
```bash
echo "install pythia8" | python3 /path/to/bin/mg5_aMC
```
returns an exit status `0`, even if [something goes wrong](https://github.com/madminer-tool/madminer-jupyter-env/issues/8#issuecomment-2268419526) during the installation step.

One possible solution I can think of is to check installation log files (using tools like `grep`) or paths in `mg5_configuration.txt`, or check the existence of executables/libraries, like [this code](https://github.com/tueda/madminer-jupyter-env/commit/fb5a03218354df3069cb3dec8299b677e76849f8). It would be preferable if there were a universal way (maybe one already exists?).

**Comment by oliviermattelaer** [comment, marked as answer]
I do not know,

Since madgraph is designed as an interactive command interface,
we discourage the usage of piping (for scripting we do have command file which have handling for "confirmation" question that can sometimes pop up and kill piping setup.
(like in your case "pythia8 already installed do you want to remove previous installation?")

Maybe one think to test is to use the command
set crash_on_error True
which deactivate all the layer of the interactive command interface that give back the prompt to the user in case of a crash.
This might be enough in this case (but not clear since the crash occurs here outside of MG5aMC)

Cheers,

Olivier


On 15 Aug 2024, at 05:08, Takahiro Ueda ***@***.***> wrote:



Is there an easy way to programmatically check if a command has succeeded or not?

Unfortunately, in shell scripts,

echo "install pythia8" | python3 /path/to/bin/mg5_aMC

returns an exit status 0, even if something goes wrong<https://github.com/madminer-tool/madminer-jupyter-env/issues/8#issuecomment-2268419526> during the installation step.

One possible solution I can think of is to check installation log files (using tools like grep) or paths in mg5_configuration.txt, or check the existence of executables/libraries, like this code<https://github.com/tueda/madminer-jupyter-env/commit/fb5a03218354df3069cb3dec8299b677e76849f8>. It would be preferable if there were a universal way (maybe one already exists?).

**Reply by tueda** [reply]
Thank you for your response. I was not aware of `crash_on_error`. It's good to know (so I'll mark this as the answer).

Unfortunately, it does not work for my case. While using `crash_on_error` results in `mg5_aMC` stopping, it still returns an exit status of `0`.

This behaviour can be confirmed by using the [`madminertool/madminer-jupyter-env:0.3.7` Docker image](https://hub.docker.com/layers/madminertool/madminer-jupyter-env/0.3.7/images/sha256-5394ae688046df93dd29e42537cf14738acf8b42ae513586d2a77a174c6b9ae3):
```bash
docker run -it --rm madminertool/madminer-jupyter-env:0.3.7 bash
```
Run `mg5_aMC` in the container:
```bash
/madminer/software/MG5_aMC_v2_9_16/bin/mg5_aMC
```
Skip the update, turn `crash_on_error` on, and try to install `mg5amc_py8_interface`:
```
n
set crash_on_error T
install mg5amc_py8_interface
```
The last `install` command fails due to the ["pthread" issue](https://gitlab.com/Pythia8/releases/-/issues/462). `mg5_aMC` terminates but with an exit status of `0`:
```bash
echo $?
```
which prints `0`.

**Comment by oliviermattelaer** [comment]
I do not think that the issue is here in MG5 side:

If I'm forcing MG5 to crash with forcing the installer to return "2" as code output, then I do have the following (with crash_on_error=T):
```
DEBUG:  return_code =  2 [madgraph_interface.py at line 6168]
Command "install mg5amc_py8_interface " interrupted in sub-command:
"install mg5amc_py8_interface" with error:
InvalidCmd : Installation of mg5amc_py8_interface failed.
Please report this bug on https://bugs.launchpad.net/mg5amcnlo
More information is found in 'MG5_debug'.
Please attach this file to your report.
stop computation due to crash_on_error=True
Command "install mg5amc_py8_interface " interrupted in sub-command:
"install mg5amc_py8_interface" with error:
InvalidCmd : Installation of mg5amc_py8_interface failed.
Please report this bug on https://bugs.launchpad.net/mg5amcnlo
More information is found in 'MG5_debug'.
Please attach this file to your report.
stop computation due to crash_on_error=True
Traceback (most recent call last):
  File "/Users/omattelaer/Documents/git_workspace/version3/./bin/mg5_aMC", line 218, in <module>
    cmd_line.cmdloop()
  File "/Users/omattelaer/Documents/git_workspace/version3/madgraph/interface/extended_cmd.py", line 969, in cmdloop
    self.error_handling(error, line)
  File "/Users/omattelaer/Documents/git_workspace/version3/madgraph/interface/extended_cmd.py", line 1509, in error_handling
    stop = self.nice_error_handling(error, line)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/omattelaer/Documents/git_workspace/version3/madgraph/interface/extended_cmd.py", line 967, in cmdloop
    stop = self.onecmd(line)
           ^^^^^^^^^^^^^^^^^
  File "/Users/omattelaer/Documents/git_workspace/version3/madgraph/interface/extended_cmd.py", line 1548, in onecmd
    return self.error_handling(error, line)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/omattelaer/Documents/git_workspace/version3/madgraph/interface/extended_cmd.py", line 1509, in error_handling
    stop = self.nice_error_handling(error, line)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/omattelaer/Documents/git_workspace/version3/madgraph/interface/extended_cmd.py", line 1546, in onecmd
    return self.onecmd_orig(line, **opt)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/omattelaer/Documents/git_workspace/version3/madgraph/interface/extended_cmd.py", line 1495, in onecmd_orig
    return func(arg, **opt)
           ^^^^^^^^^^^^^^^^
  File "/Users/omattelaer/Documents/git_workspace/version3/madgraph/interface/master_interface.py", line 490, in do_install
    self.cmd.do_install(self, *args, **opts)
  File "/Users/omattelaer/Documents/git_workspace/version3/madgraph/interface/madgraph_interface.py", line 6514, in do_install
    return self.advanced_install(name, path['HEPToolsInstaller'],
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/omattelaer/Documents/git_workspace/version3/madgraph/interface/madgraph_interface.py", line 6209, in advanced_install
    raise self.InvalidCmd("Installation of %s failed."%tool_to_install)
madgraph.InvalidCmd: Installation of mg5amc_py8_interface failed.
[version3]$ echo $?
1
```

So everything seems to works as I would expect on the MG5aMC side (which is good news since the installer can be updated without touching MG5aMC code).
So now the issue can be two-sided:
1) from the HEPToolsInstaller side (not catching correctly an issue in the compilation step)
2) in the code to be installed itself (not reporting the crash to HEPToolsInstaller)

I have checked #1 and indeed the return code value is just ignored for other types of check for successful installation.
This is obviously a big change in HEPToolsInstaller since all the installation function need to be modified accordingly.
https://github.com/mg5amcnlo/HEPToolsInstallers/commit/7d9e3b6bfcc491e371124a6df630ff9bbdf5c08d

Which likely is not enough for all tools, but likely enough for this interface.  Other codes often use an intermediate bash script that can also catch error code from make/cmake and not report it back. But we will need to handle that when we have an example where this occurs).

The CI should update the install command at 11h30 (so in 13 mins) so you should be able to test (the second point) within your setup soon (likely when you read this notification).

Cheers,

Olivier

**Reply by tueda** [reply]
Thank you for the update.

I am testing MG5_aMC v2.9.16 in the Docker image and so its behaviour might be different from your development version on the `3.x` branch, but if I run it in debug mode (`mg5_aMC --debug`), the `crash_on_error` results in `stop computation due to crash_on_error=True`, prints the traceback and returns an exit status of `1`. Without the `--debug` flag, it stops (without printing the traceback) but returns an exit status of `0`.

Is this difference in the exit statuses with and without the `--debug` flag intended?
